psh/cp/cp.c

```diff
@@ -0,0 +1,185 @@
+
+#include "../psh.h"
+
+#define SIZE_BUFF 256
+
+
+static void psh_cpinfo(void)
+{
+	printf("copy file");
+}
+
+
+static void psh_cp_help(const char *prog)
+{
+	printf("Usage: %s [options] SOURCE TARGET\n", prog);
+	printf("  -h:  shows this help message\n");
+}
+
+
+static int psh_cp(int argc, char **argv)
+{
+	FILE *src = NULL, *dst = NULL;
+	int fdsrc, fddst;
+	unsigned char buff[SIZE_BUFF];
+	char *filename, *destpath;
+	size_t destlen, filelen, cnt;
+	struct stat stat;
+	int c, retval = EXIT_SUCCESS, eexit = 0;
+
+	while ((c = getopt(argc, argv, "h")) != -1) {
+		switch (c) {
+			case 'h':
+				psh_cp_help(argv[0]);
+				eexit = 1;
+				break;
+			default:
+				psh_cp_help(argv[0]);
+				retval = EXIT_FAILURE;
+				eexit = 1;
+				break;
+		}
+	}
+
+	if (argc - optind != 2) {
+		psh_cp_help(argv[0]);
+		retval = EXIT_FAILURE;
+		eexit = 1;
+	}
+
+	while (eexit == 0) {
+		fdsrc = open(argv[optind], O_RDONLY);
+		if (fdsrc < 0) {
+			retval = EXIT_FAILURE;
+			fprintf(stderr, "cp: could not open source file\n");
+			break;
+		}
+
+		if (fstat(fdsrc, &stat) < 0) {
+			close(fdsrc);
+			retval = EXIT_FAILURE;
+			fprintf(stderr, "cp: stat failed\n");
+		}
+
+		if (S_ISDIR(stat.st_mode) != 0) {
+			close(fdsrc);
+			retval = EXIT_FAILURE;
+			fprintf(stderr, "cp: could not open source file - is a directory\n");
+			break;
+		}
+
+		src = fdopen(fdsrc, "r");
+		if (src == NULL) {
+			close(fdsrc);
+			retval = EXIT_FAILURE;
+			fprintf(stderr, "cp: could not open source file\n");
+			break;
+		}
+
+		fddst = open(argv[optind + 1], O_WRONLY | O_CREAT, DEFFILEMODE);
+		if (fddst < 0) {
+			if (errno == EISDIR) {
+				close(fddst);
+
+				filename = strrchr(argv[optind], '/');
+				if (filename == NULL) {
+					filename = argv[optind];
+				}
+
+				destlen = strlen(argv[optind + 1]);
+				filelen = strlen(filename);
+				destpath = malloc(destlen + filelen + 2);
+				if (destpath == NULL) {
+					retval = EXIT_FAILURE;
+					fprintf(stderr, "cp: out of memory\n");
+					break;
+				}
+
+				strncpy(destpath, argv[optind + 1], destlen);
+				destpath[destlen] = '/';
+				strcpy(destpath + destlen + 1, filename);
+
+				fddst = open(destpath, O_WRONLY | O_CREAT, DEFFILEMODE);
+				free(destpath);
+				if (fddst < 0) {
+					retval = EXIT_FAILURE;
+					fprintf(stderr, "cp: could not open destination file\n");
+					break;
+				}
+			}
+			else {
+				close(fddst);
+				fprintf(stderr, "cp: could not open destination file\n");
+				retval = EXIT_FAILURE;
+				break;
+			}
+		}
+
+		dst = fdopen(fddst, "w");
+		if (dst == NULL) {
+			close(fddst);
+			retval = EXIT_FAILURE;
+			fprintf(stderr, "cp: could not open destination file\n");
+			break;
+		}
+
+		while (1) {
+			cnt = fread(buff, sizeof(unsigned char), sizeof(buff), src);
+			if (cnt == 0) {
+				if (ferror(src) != 0) {
+					retval = EXIT_FAILURE;
+					fprintf(stderr, "cp: read failure\n");
+				}
+				break;
+			}
+
+			cnt = fwrite(buff, sizeof(unsigned char), cnt, dst);
```

Comment:
probably went with `fread`, `fwrite` to avoid partial writes, but take care of possible double buffering.

Right now `fwrite` will be double-buffered on MMU targets as `SIZE_BUFF` is smaller than `BUFSIZ`. With `fread` the situation is similar (altrough on MMU it will cache 4096 bytes, on NOMMU it will cache first 16 bytes, copy them, then read remaining 256-16 bytes directly).

Maybe we should check `BUFSIZ` (exported by libc) and set `SIZE_BUFF` accordingly - or explicitly remove buffering from these streams?

consider using `fread_unlocked` and `fwrite_unlocked` to speed up the copying.

Reply:
Done